nova/virt/vmwareapi/driver.py

```diff
@@ -690,6 +697,62 @@ def detach_interface(self, context, instance, vif):
         """Detach an interface from the instance."""
         self._vmops.detach_interface(context, instance, vif)
 
+    def sync_server_group(self, context, sg_uuid):
+        self._vmops.sync_server_group(context, sg_uuid)
+
+    def _server_group_sync_loop(self, compute_host):
+        """Retrieve all groups from the cluster and from the DB and call
+        self.sync_server_group() for them.
+
+        We always wait a little not not overwhelm the cluster.
+        """
+        context = nova_context.get_admin_context()
+
+        while CONF.vmware.server_group_sync_loop_spacing >= 0:
+            LOG.debug('Starting server-group sync-loop')
+            try:
+
+                InstanceList = objects.instance.InstanceList
+                instance_uuids = set(InstanceList.get_by_host(
+                    context, compute_host, expected_attrs=[]))
+
+                InstanceGroupList = objects.instance_group.InstanceGroupList
+                ig_list = InstanceGroupList.get_by_instance_uuids(
+                    context, instance_uuids)
+
+                sg_uuids = set(ig.uuid for ig in ig_list)
+
+                cluster_rules = cluster_util.fetch_cluster_rules(
+                    self._session, cluster_ref=self._cluster_ref)
+                for rule in cluster_rules:
+                    if not rule.name.startswith(constants.DRS_PREFIX):
+                        # not managed by us
+                        continue
+
+                    # should look like
+                    # NOVA_d7134f26-f2e2-42ed-b5f1-4092962e84d5-affinity
+                    # NOVA_d7134f26-f2e2-42ed-b5f1-4092962e84d5-soft-anti-affinity
+                    # NOVA_d7134f26-f2e2-42ed-b5f1-4092962e84d5-soft-affinity
+                    m = UUID_RE.search(rule.name)
+                    if not m:
+                        continue
+
+                    sg_uuids.add(m.group(0))
+
+                for sg_uuid in sg_uuids:
+                    spacing = \
+                        CONF.vmware.server_group_sync_loop_max_group_spacing
+                    sleep_time = random.uniform(0.5, spacing)
+                    time.sleep(sleep_time)
+                    self._vmops.sync_server_group(context, sg_uuid)
+            except Exception as e:
+                LOG.exception("Finished server-group sync-loop with error: %s",
+                              e)
```

Comment:
Also not critical, but I think, you are duplicating here the exception in the output.
At least the standard logger from python logs with exception in the last line
```ipython
In [1]: import logging
   ...: LOG = logging.getLogger("mylogger")
   ...: def a():
   ...:     try:
   ...:         1/0
   ...:     except Exception as e:
   ...:         LOG.exception("a %s", e)
   ...:

In [2]: a()
a division by zero
Traceback (most recent call last):
  File "<ipython-input-1-2b4dff5ef495>", line 5, in a
    1/0
ZeroDivisionError: division by zero```

Reply:
True, but if I search for "Finished server-group sync-loop" this will show the error in the same line without me having to invent some "and please also the line after if there's a "with error" in my line" magic, that I don't know how to do.